docs/proposals/odo-config-file.md

```diff
@@ -0,0 +1,253 @@
+1. As a user, I want to be able to use configuration stored with my source code to help me be more productive
+1. As a user, I want to be able to manage explicit configuration that is saved with my source code
+1. As a user, I want to be able to check out an application source code on different computers and have them producing the same result in relation to the development environment (deployment of the application on the platform)
+
+## Design overview
+
+We need to provide a configuration file `odo-config` in a configuration directory `.odo`, that will hold specific configuration that `odo` will use on execution. This configuration might modify how the CLI work or how the component will work on the platform.
+
+### Scopes of configuration
+
+`odo` CLI configuration needs to be preserved in configuration files. As a developer might work on multiple
+components, there is a need to have different scopes for configuration so that a developer don't need to
+explicitly configure each and every component independently.
+
+#### Local configuration
+
+Every component can be configured independently. We call this local configuration for the component as it only affects the component it references. This configuration will be stored in a file in `.odo` specific directory located at the root directory of the component's source code. In this way, the configuration file can be added to version control system and preserved along with the application. If any other developer checks out the same source code he will get the same configuration as the original developer.
+
+We propose to have a `.odo` folder in the root of the application's source code directory to store every odo local configurations.
+
+```bash
+<app-dir>/.odo/odo-config
+```
+
+A good representation of this is a developer that while on the development process realizes his application requires 2 GB of memory to run appropriately, he might decide to save that information in this configuration file. At a later moment, any other developer (or even him in the future) that decides to work on the same application will have this configuration definition along with his source code. `odo` should take into account this configuration definition and materialize when appropriate. The type of configuration, format and applicability will be described later in this document.
+
+#### Global configuration
+
+When working with multiple components it might be tedious to configure each and every component, or a user wants to have a default configuration value applied to each and every component he works with when no explicit (local) configuration exists. For this purpose, there should be a file, stored in the user's home directory's odo specific folder, that should contain this configuration definitions. This configuration is global to every component the user works with.
+
+We propose to have a `.odo` folder in the root of the users home directory to store every odo global configurations.
```

Comment:
also do we require `.odo` for global?

Reply:
Same as before. Not so much currently needed, but this might be a place where maybe we store plugins, or plugin configuration, so yes, better to have a tool folder.

Reply:
sounds good, thanks for clarifying :thumbsup: